### ISSUE TITLE
Feature Request: solc 0.8 and solc 0.7

### ISSUE THREAD

**Issue by gitpusha** [ethereum/solidity#10317]
I guess this is more like a santa wish list. I did not by any means research the complexity of the underlying issues. I am just getting it out there for santa 🎅 

1. full support for `calldata` arrays / structs (see https://github.com/ethereum/solidity/issues/9160)

2. `try catch` for `abi.decode` 

3. `try catch` for internal/private functions (I guess not needed if `try catch` for failing things like e.g.`abi.decode` exists )

4. `array slices` for `memory` arrays 

5. a nicer way to convert `bytes32` to `address` other than 

```solidity
address(uint160(uint256(_someWord)));
```

6. make it so that when batch `delegatecalling` targets from a contract, not all of them have to be `payable` functions in order for there to be a `non-zero` `msg.value` property (probably impossible to implement due to EVM restriction? ) 

7. file-level events 

8. file-level enums

**Comment by chriseth** [comment]
Thank you for opening this issue! We are having trouble prioritizing features, so this in very valuable to us! Most of the requests are non-breaking and thus not relevant for the actual 0.8 release.

Most of the issues are high on our backlog, so I won't comment on them.

2.: try/catch is designed for external calls, but I can see the problem with `abi.decode`. Instead of try/catch we should maybe rather have `abi.tryDecode`? The problem here is how to return the data in the two cases. Can you maybe create a new issue for that, so we can discuss it in detail?

3.: Same thing - which other functions would you consider? (Please answer in the new issue)

5.: This is relevant for 0.8 because we actually want to restrict explicit conversions even more. I currently don't see a better way. Do you do this conversion often? 

6.: Yes, this is an EVM restriction. Solidity library functions actually do not have a check for msg.value for exactly this purpose. Maybe we could find another way to specify "this is a delegatecall target function"? Or maybe we should modify libraries so that they can be more easily used like that? Can you give an example (maybe in a new issue)?

**Comment by gitpusha** [comment]
No problem, I appreciate your round the clock communication on language support. 

2.: `abi.tryDecode`:  I opened an issue #10381  to continue the discussion.

3.: will report in the issue but no other examples yet.

5.: No, I don't do this conversion very often. Just when I do it it looks cryptic to other people reading the code and me. So would still be nice to have. I might make more use of this conversion in the future. 

6.: I opened an issue here: #10382
